filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobFileSystemManager.scala

```diff
@@ -169,54 +179,81 @@ object BlobSasTokenGenerator {
     * Else try to use the environment azure identity to request the SAS token
     *
     * @return an AzureSasCredential for accessing a blob container
     */
-  def generateBlobSasToken: Try[AzureSasCredential] = {
+  def findBlobSasToken(endpoint: EndpointURL, container: BlobContainerName, buffer: Duration): Try[AzureSasCredential] = {
+     getAvailableCachedSasToken(endpoint, container) match {
+      case Some(sas) if BlobFileSystemManager.isSasValid(sas, buffer) => Success(sas)
+      // If unavailable or expired refresh SAS cache entry
+      case _ => {
+        val azureSasTokenTry: Try[AzureSasCredential] = generateBlobSasToken(endpoint, container)
+        azureSasTokenTry match {
+          case Success(sas) => putAvailableCachedSasToken(endpoint, container, sas)
+          case _ => ()
+        }
+        azureSasTokenTry
+      }
+    }
+  }
+
+  /**
+    * Requests new SAS token from WSM client for a specified endpoint and container
+    *
+    * @return a new AzureSasCredential for accessing a blob container
+    */
+  def generateBlobSasToken(endpoint: EndpointURL, container: BlobContainerName): Try[AzureSasCredential] = {
     val wsmAuthToken: Try[String] = overrideWsmAuthToken match {
       case Some(t) => Success(t)
       case None => AzureCredentials.getAccessToken(None).toTry
     }
-
     for {
       wsmAuth <- wsmAuthToken
-      wsmClient = wsmClientProvider.getControlledAzureResourceApi(wsmAuth)
-      sasToken <- Try(  // Java library throws
-        wsmClient.createAzureStorageContainerSasToken(
-          workspaceId.value,
-          containerResourceId.value,
-          null,
-          null,
-          null,
-          null
-        ).getToken)
-    } yield new AzureSasCredential(sasToken)
+      wsmAzureResourceClient = wsmClientProvider.getControlledAzureResourceApi(wsmAuth)
+      sa <- endpoint.storageAccountName
+      ids <- getWorkspaceAndContainerResourceId(sa, container, wsmAuth)
+      azureSasToken <- wsmAzureResourceClient.createAzureStorageContainerSasToken(ids._1, ids._2)
+    } yield azureSasToken
+  }
+
+  def getWorkspaceAndContainerResourceId(storageAccountName: StorageAccountName, container: BlobContainerName, wsmAuth : String): Try[(UUID, UUID)] = {
+    val wsmWorkspaceClient = wsmClientProvider.getWorkspaceApi(wsmAuth)
+    val wsmResourceClient = wsmClientProvider.getResourceApi(wsmAuth)
+    val wsmLandingZonesClient = wsmClientProvider.getLandingZonesApi(wsmAuth)
+    for {
+      workspaceId <- container.workspaceId
+      billingProfile <- wsmWorkspaceClient.findBillingProfileId(workspaceId)
+      landingZone <- wsmLandingZonesClient.findLandingZoneForBillingProfile(billingProfile)
+      storageAccountId <- wsmLandingZonesClient.findStorageAccountForLandingZone(landingZone)
+      resourceId <- wsmResourceClient.findContainerResourceId(workspaceId, container)
+      if (storageAccountId.value.contains(storageAccountName.value))
```

Comment:
As we discussed earlier - I think doing it this way was my idea in the first place (sorry 😬 ) but I'm not sure it's worth matching the storage account. In the unlikely event that doing so saved us from some mismatch, the result would be an error and not inappropriate access.